src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,192 @@
+            bool includesCommitCharacter;
+            if (!subjectBuffer.CheckEditAccess())
+            {
+                // We are on the wrong thread.
+                FatalError.ReportWithoutCrash(new InvalidOperationException("Subject buffer did not provide Edit Access"));
+                return AsyncCompletionData.CommitBehavior.None;
+            }
+
+            if (subjectBuffer.EditInProgress)
+            {
+                FatalError.ReportWithoutCrash(new InvalidOperationException("Subject buffer is editing by someone else."));
+                return AsyncCompletionData.CommitBehavior.None;
+            }
+
+            using (var edit = subjectBuffer.CreateEdit())
+            {
+                var change = completionService.GetChangeAsync(document, roslynItem, commitCharacter, cancellationToken).WaitAndGetResult(cancellationToken);
+
+                var textChange = change.TextChange;
+
+                var triggerSnapshotSpan = new SnapshotSpan(triggerSnapshot, textChange.Span.ToSpan());
+                var mappedSpan = triggerSnapshotSpan.TranslateTo(subjectBuffer.CurrentSnapshot, SpanTrackingMode.EdgeInclusive);
+
+                edit.Replace(mappedSpan.Span, change.TextChange.NewText);
+                edit.Apply();
+
+                if (change.NewPosition.HasValue)
+                {
+                    view.TryMoveCaretToAndEnsureVisible(new SnapshotPoint(subjectBuffer.CurrentSnapshot, change.NewPosition.Value));
+                }
+
+                includesCommitCharacter = change.IncludesCommitCharacter;
+
+                if (roslynItem.Rules.FormatOnCommit)
+                {
+                    // refresh the document
+                    document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                    var spanToFormat = triggerSnapshotSpan.TranslateTo(subjectBuffer.CurrentSnapshot, SpanTrackingMode.EdgeInclusive);
+                    var formattingService = document?.GetLanguageService<IEditorFormattingService>();
+
+                    if (formattingService != null)
+                    {
+                        var changes = formattingService.GetFormattingChangesAsync(
+                            document, spanToFormat.Span.ToTextSpan(), CancellationToken.None).WaitAndGetResult(CancellationToken.None);
+                        document.Project.Solution.Workspace.ApplyTextChanges(document.Id, changes, CancellationToken.None);
+                    }
+                }
+            }
+
+            if (includesCommitCharacter)
+            {
+                return AsyncCompletionData.CommitBehavior.SuppressFurtherTypeCharCommandHandlers;
+            }
+
+            if (commitCharacter == '\n' && Controller.SendEnterThroughToEditor(rules, roslynItem, filterText))
+            {
+                return AsyncCompletionData.CommitBehavior.RaiseFurtherReturnKeyAndTabKeyCommandHandlers;
+            }
+
+            return AsyncCompletionData.CommitBehavior.None;
```

Comment:
this seems to have changed a bunch from Controller_Commit.  It's hard to tell if things have stayed the same.  For example, where did the ComputeMinimalChange code go?  Where's the code to SetCaretPosition?

Reply:
As I see, this code is on the Editor side now: AsyncCompletionSession.UpdateCaretPosition.
@amadeusW to confirm

---
In reply to: [232138203](https://github.com/dotnet/roslyn/pull/29016#discussion_r232138203) [](ancestors = 232138203)

Reply:
1. The editor team already merged some PRs on their side. Their code is available here: [https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=%2Fsrc%2FEditor%2FLanguage%2FImpl%2FLanguage%2FAsyncCompletion&version=GBmaster](https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=%2Fsrc%2FEditor%2FLanguage%2FImpl%2FLanguage%2FAsyncCompletion&version=GBmaster)

@cyrusnajmabadi, I do not know if you have an access to it. @jinujoseph, I do not know if we are able to provide an access to VS-Platform repo to @cyrusnajmabadi

2. We provide some information about the caret via CommitBehavior. Some information is completely handled on the Editor side. @amadeusW can provide more information.

---
In reply to: [232429491](https://github.com/dotnet/roslyn/pull/29016#discussion_r232429491) [](ancestors = 232429491)

Reply:
yeah.  Unfortunately, i can't reach it.  That said, what i'm looking for is this part:

> could you point in this PR to where this is communicated? How are we passing the information to them on where the caret should go? Thanks! #Resolved

Can you answer that bit?

Reply:
1. Session subscribes to 
_textView.Caret.PositionChanged
2. If the event happens, they asynchronously UpdateCaretPosition.

Please let me know if you have more specific questions.

---
In reply to: [232835185](https://github.com/dotnet/roslyn/pull/29016#discussion_r232835185) [](ancestors = 232835185)

Reply:
Sorry: I'm not being clear:

Completion providers has the ability to specify where they want the care to end up.  I"m asking: where is the code in that new system that passes this information along so that post-commit hte caret is where it is supposed to be.

Reply:
OK. Let us use broker.CompletionTriggered. @amadeusw to review. Thank you!

---
In reply to: [233249517](https://github.com/dotnet/roslyn/pull/29016#discussion_r233249517) [](ancestors = 233249517)

Reply:
Removed events. Using RecentItemsManager shared between CommitManager and ItemManager.

---
In reply to: [233279063](https://github.com/dotnet/roslyn/pull/29016#discussion_r233279063) [](ancestors = 233279063,233249517)